crates/re_viewer_context/src/space_view/space_view_class.rs

```diff
@@ -180,184 +213,55 @@ pub trait SpaceViewClass: std::marker::Sized + Send + Sync {
-        space_origin: &EntityPath,
-        space_view_id: SpaceViewId,
-        root_entity_properties: &mut EntityProperties,
-    ) {
-        typed_state_wrapper_mut(state, |state| {
-            self.selection_ui(
-                ctx,
-                ui,
-                state,
-                space_origin,
-                space_view_id,
-                root_entity_properties,
-            );
-        });
+/// Implementation of an empty space view state.
+impl SpaceViewState for () {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
     }
 
-    #[allow(clippy::for_kv_map)]
-    fn ui(
-        &self,
-        ctx: &ViewerContext<'_>,
-        ui: &mut egui::Ui,
-        state: &mut dyn SpaceViewState,
-        root_entity_properties: &EntityProperties,
-        query: &ViewQuery<'_>,
-        system_output: SystemExecutionOutput,
-    ) {
-        re_tracing::profile_function!();
-
-        typed_state_wrapper_mut(state, |state| {
-            if let Err(err) = self.ui(ctx, ui, state, root_entity_properties, query, system_output)
-            {
-                // TODO(andreas): Draw an error message on top of the space view ui instead of logging.
-                re_log::error_once!("Error drawing ui for space view: {err}");
-            }
-        });
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
     }
 }
 
-fn typed_state_wrapper_mut<T: SpaceViewState, R: Default, F: FnOnce(&mut T) -> R>(
-    state: &mut dyn SpaceViewState,
-    fun: F,
-) -> R {
-    if let Some(state) = state.as_any_mut().downcast_mut() {
-        fun(state)
-    } else {
-        re_log::error_once!(
-            "Unexpected space view state type. Expected {}",
-            std::any::type_name::<T>()
-        );
-        R::default()
+pub trait SpaceViewStateExt: SpaceViewState {
+    /// Downcasts this state to a reference of a concrete type.
+    fn downcast_ref<T: SpaceViewState>(&self) -> Result<&T, SpaceViewSystemExecutionError> {
+        self.as_any()
+            .downcast_ref()
+            .ok_or(SpaceViewSystemExecutionError::StateCastError(
+                std::any::type_name::<T>(),
+            ))
     }
-}
 
-fn typed_state_wrapper<T: SpaceViewState, R: Default, F: FnOnce(&T) -> R>(
-    state: &dyn SpaceViewState,
-    fun: F,
-) -> R {
-    if let Some(state) = state.as_any().downcast_ref() {
-        fun(state)
-    } else {
-        re_log::error_once!(
-            "Unexpected space view state type. Expected {}",
-            std::any::type_name::<T>()
-        );
-        R::default()
+    /// Downcasts this state to a mutable reference of a concrete type.
+    fn downcast_mut<T: SpaceViewState>(&mut self) -> Result<&mut T, SpaceViewSystemExecutionError> {
+        self.as_any_mut()
+            .downcast_mut()
+            .ok_or(SpaceViewSystemExecutionError::StateCastError(
+                std::any::type_name::<T>(),
+            ))
     }
 }
+
+impl SpaceViewStateExt for dyn SpaceViewState {}
```

Comment:
As per discussion in slack, that might be just ignorance, but I'm wondering if all of that boilerplate could be avoided by using `dyn Any` instead of using a `SpaceViewState` trait at all in the `SpaceViewClass` functions.

Reply:
played around with it a bit more, there's a bunch of drawbacks:
* loose all constraints on what type `new_state` returns and the other methods get
* can't implement the utilitites that downcast + convert to `Result<.., SpaceViewSystemExecutionError>` which is nice because then we can just `?` on most of the casts, saving us quite a bit of code

`as_any` and `as_any_mut` are required because they are the only way to get from an arbitrary trait pointer to a `dyn Any` which is what has the "inner" downcast_ref/mut methods we need here. I found this blog post helpful in explaining it https://lucumr.pocoo.org/2022/1/7/as-any-hack/
(Fairly sure I also saw an RFC on Rust that would make this obsolete somewhere)

We might want to revisit if we instantiate `SpaceViewClass` per space view instance and allow storing state in concrete `SpaceViewClass` impls, but I'm on very much on the fence about that: most functions are strictly about the class and not any particular instance and I don't want implementors to be able to e.g. change the display name of _the class_ based on internal state